Frameworks/ImageIO/CGImageSource.mm

```diff
@@ -489,21 +1370,109 @@ size_t CGImageSourceGetCount(CGImageSourceRef isrc) {
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Caveat
+ @Notes A limited subset of source container properties are supported currently  
 */
 CFDictionaryRef CGImageSourceCopyProperties(CGImageSourceRef isrc, CFDictionaryRef options) {
-    UNIMPLEMENTED();
-    return StubReturn();
+    RETURN_NULL_IF(!isrc);
+    NSData* imageData = ((ImageSource*)isrc).data;
+    RETURN_NULL_IF(!imageData);
+
+    NSMutableDictionary *properties = [[NSMutableDictionary alloc] init]; 
+    [properties setObject:[NSNumber numberWithInt:[imageData length]] forKey:(id)kCGImagePropertyFileSize];
+
+    // Look for container-level metadata if the file is a GIF.
+    CFStringRef imageFormat = CGImageSourceGetType(isrc);
+    if (imageFormat == kUTTypeGIF) {
+        MULTI_QI imageQueryInterface = {0};
+        const GUID IID_IWICImagingFactory = {0xec5ec8a9,0xc395,0x4314,0x9c,0x77,0x54,0xd7,0xa9,0x35,0xff,0x70};
+        imageQueryInterface.pIID = &IID_IWICImagingFactory;
+        RETURN_NULL_IF_FAILED(
+            CoCreateInstanceFromApp(CLSID_WICImagingFactory, nullptr, CLSCTX_INPROC_SERVER, nullptr, 1, &imageQueryInterface));
+
+        ComPtr<IWICImagingFactory> imageFactory = (IWICImagingFactory*)imageQueryInterface.pItf;
+        ComPtr<IWICStream> imageStream;
+        RETURN_NULL_IF_FAILED(imageFactory->CreateStream(&imageStream));
+
+        unsigned char* imageByteArray = (unsigned char*)[imageData bytes];
+        RETURN_NULL_IF_FAILED(imageStream->InitializeFromMemory(imageByteArray, [imageData length]));
+
+        if (options && CFDictionaryContainsKey(options, kCGImageSourceShouldCache)) {
+            UNIMPLEMENTED_WITH_MSG("kCGImageSourceShouldCache is not supported in current implementation.");
+        }
+
+        ComPtr<IWICBitmapDecoder> imageDecoder;                
+        RETURN_NULL_IF_FAILED(
+            imageFactory->CreateDecoderFromStream(imageStream.Get(), nullptr, WICDecodeMetadataCacheOnDemand, &imageDecoder));    
+        ComPtr<IWICMetadataQueryReader> imageMetadataReader;
+        RETURN_NULL_IF_FAILED(imageDecoder->GetMetadataQueryReader(&imageMetadataReader));
+
+        PROPVARIANT propertyValue;
+        PropVariantInit(&propertyValue);
+        NSMutableDictionary *gifProperties = [[NSMutableDictionary alloc] init];
+
+        if (SUCCEEDED(imageMetadataReader->GetMetadataByName(L"/appext/Data", &propertyValue)) &&
+            propertyValue.vt == (VT_VECTOR | VT_UI1)) {
+            if (propertyValue.caub.pElems[0] == 3 && propertyValue.caub.pElems[1] == 1) {
+                NSNumber* loopCount = [NSNumber numberWithInt:propertyValue.caub.pElems[2] + (propertyValue.caub.pElems[3]<<8)];
+                [gifProperties setObject:loopCount forKey:(id)kCGImagePropertyGIFLoopCount];
+            }
```

Comment:
Added a comment about what the GIF header contains for the looping properties, and made constants for the static bytes in front.